be/btree.c

```diff
@@ -1256,17 +1256,16 @@ static void btree_truncate(struct m0_be_btree *btree, struct m0_be_tx *tx,
 			 * node->bt_child_arr[node->bt_num_active_key]
 			 * leaf node
 			 */
-			if (i > 0 && limit > 0) {
+			if (i > 0)
 				btree_pair_release(btree, tx,
 						&parent->bt_kv_arr[i-1]);
 
+			if (limit > 0)
 				limit--;
-				parent->bt_num_active_key--;
-			}
-
 			if (i == 0)
 				parent->bt_isleaf = true;
-
+			else
+				parent->bt_num_active_key--;
```

Comment:
1. `i` should be defined as "unsigned int", the same as struct m0_be_bnode::bt_num_active_key.

2. The checking of `i` in these two places can be combined.
```
			if (i > 0) {
				btree_pair_release(btree, tx,
						&parent->bt_kv_arr[i-1]);
                               parent->bt_num_active_key--;
                       } else {
                               /* now i must be 0. */
                               parent->bt_isleaf = true;
                       }

			if (limit > 0)
				limit--;				
```

Reply:
@huanghua78 I will accept the first suggestion to make i as 'unsigned int'. The second suggestion while logically correct I will skip because:
1) The code this way has been working for years and I am worried of making changes and not covering any path that might fail (though I agree there does not seem to be any such unknown path).
2) The optimizing compiler will anyway club the statements as you mentioned thus avoiding multiple checks for 'i' and achieving the same effect as you have put in your code snippet.

Reply:
Agreed.